src/main/java/com/google/sps/utils/ServletUtils.java

```diff
@@ -55,12 +61,57 @@ public static void printBadRequestError(HttpServletResponse response, String err
     response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
     response.getWriter().println(errorMessage);
   }
-  
+
   public static void enforceUserLogin(HttpServletRequest request, HttpServletResponse response) throws IOException {
     if(request.getUserPrincipal() == null) {
       // If user is not logged in, redirect to index page.
       response.sendRedirect("/");
       return;
     }
   }
+
+  public static ImmutableList<ImmutableMap<String, String>> getGroupsList(HttpServletRequest request) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    String userEmail = request.getUserPrincipal().getName();
+    // Stores the IDs for all the groups the user ahs joined
+    HashSet<Long> userGroups = new HashSet<Long>();
+
+    Query userQuery = new Query(USER_KIND).addFilter(Entity.KEY_RESERVED_PROPERTY, FilterOperator.EQUAL, userEmail);
+    PreparedQuery userPreparedQuery = datastore.prepare(userQuery);
+    Entity user = userPreparedQuery.asSingleEntity();
+
+    if (user != null) {
+      userGroups = (HashSet<Long>) user.getProperty(GROUPS_KEY);
+
+    } else {
+      // if the user does not exist, create one and add to datastore
+      user = new Entity(USER_KIND, userEmail);
+
+      user.setProperty(USER_EMAIL_PROPERTY, userEmail);
+      user.setProperty(GROUPS_KEY, userGroups);
+
+      datastore.put(user);
+    }
+
+    ImmutableList.Builder<ImmutableMap<String, String>> groupsListBuilder = new ImmutableList.Builder<>();
+    for (Long groupId : userGroups) {
+      try {
+        Entity group = datastore.get(KeyFactory.createKey(GROUP_KIND, groupId));
+        ImmutableMap<String, String> groupMap = ImmutableMap.of(
+            GROUP_NAME_PROPERTY, (String) group.getProperty(GROUP_NAME_PROPERTY),
+            GROUP_ID_PROPERTY, group.getKey().getName(),
+            GROUP_DESCRIPTION_PROPERTY, (String) group.getProperty(GROUP_DESCRIPTION_PROPERTY),
+            GROUP_CALENDARID_PROPERTY, (String) group.getProperty(GROUP_CALENDARID_PROPERTY),
+            GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY));
+        groupsListBuilder.add(groupMap);
+      } catch (EntityNotFoundException invalidGroup) {
+        continue;
+      }
+            
+    }
+    ImmutableList<ImmutableMap<String, String>> groupsList = groupsListBuilder.build();
+    return groupsList; 
```

Comment:
you can condense this into one line and just do `return groupsListBuilder.build();`